### PR TITLE
8256823: C2 compilation fails with "assert(isShiftCount(imm8 >> 1)) failed: illegal shift count"

### PR DESCRIPTION
**Pull request by TobiHartmann**
The ideal transformation added by [JDK-8254872](https://bugs.openjdk.java.net/browse/JDK-8254872) converts `RotateLeftNode(val, shift)` into `RotateRightNode(val, 32/64 - (shift & 31/63))`. If `shift` later becomes zero, we end up trying to emit a rotate with a 32/64 shift value which triggers an assert.

I've added an identity transformation similar to what is implemented for ShiftNodes that takes care of this. I've also noticed that the corresponding assert is not strong enough for `roll` and `rorl` (probably the author used the assert corresponding to the 64-bit version by accident). The patch also includes some refactoring.

Thanks,
Tobias

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (6/6 passed) | ❌ (2/2 failed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ⏳ (1/9 running) |    |  ⏳ (3/9 running) | ✔️ (9/9 passed) |

**Failed test tasks**
- [Linux x86 (build debug)](https://github.com/TobiHartmann/jdk/runs/1448466620)
- [Linux x86 (build release)](https://github.com/TobiHartmann/jdk/runs/1448466596)

### Issue
 * [JDK-8256823](https://bugs.openjdk.java.net/browse/JDK-8256823): C2 compilation fails with "assert(isShiftCount(imm8 >> 1)) failed: illegal shift count"

### Reviewers
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to 9c7234b08efbf50f520a91e7d7c5adf19104f00f
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 9c7234b08efbf50f520a91e7d7c5adf19104f00f
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to a207039ee0b429f904e84d8b38c8a73069a7f25e

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1384/head:pull/1384`
`$ git checkout pull/1384`

### CODE REVIEW

File: src/hotspot/share/opto/mulnode.cpp [hunk 1]
```diff
@@ -1441,6 +1441,19 @@ uint MulAddS2INode::hash() const {
 
 //------------------------------Rotate Operations ------------------------------
 
+Node* RotateLeftNode::Identity(PhaseGVN* phase) {
+  const Type* t1 = phase->type(in(1));
+  if (t1 == Type::TOP) {
+    return this;
+  }
+  // Rotate by a multiple of 32/64 does nothing
+  int mask = (t1->isa_int() ? BitsPerJavaInteger : BitsPerJavaLong) - 1;
```
Comment (iwanowww) [inline src/hotspot/share/opto/mulnode.cpp:1450]: Would be nice to assert that t1 is either TypeInt or TypeLong.

File: src/hotspot/share/opto/mulnode.cpp [hunk 2]
```diff
@@ -1441,6 +1441,19 @@ uint MulAddS2INode::hash() const {
...
+  if ((getShiftCon(phase, this, -1) & mask) == 0) {
```
Comment (iwanowww) [inline src/hotspot/share/opto/mulnode.cpp:1451]: I find the interaction between `in(2)`, `-1`, and `& mask == 0` intricate (it covers both cases: when shift constant is successfully extracted and when it is not).

Maybe refactor `getShiftCon()` to return a `bool` which signals that the constant was found and then return the value as an out argument?

**Review by iwanowww — COMMENTED** [review]
Overall, looks good. 

Some minor comments follow.

**Comment by TobiHartmann** [comment]
Thanks for the review Vladimir. I've updated the fix accordingly.

**Review by iwanowww — APPROVED** [review]
Looks good.

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by TobiHartmann** [comment]
Thanks @iwanowww and @vnkozlov for the reviews!

File: src/hotspot/share/opto/mulnode.cpp [hunk 3]
```diff
@@ -875,26 +889,28 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
...
+    // Check for useless sign-masking
+    if (in(1)->Opcode() == Op_LShiftI &&
+        in(1)->req() == 3 &&
+        in(1)->in(2) == in(2)) {
+      count &= BitsPerJavaInteger-1; // semantics of Java shifts
+      // Compute masks for which this shifting doesn't change
+      int lo = (-1 << (BitsPerJavaInteger - ((uint)count)-1)); // FFFF8000
+      int hi = ~lo;               // 00007FFF
+      const TypeInt *t11 = phase->type(in(1)->in(1))->isa_int();
+      if (!t11) return this;
```
Comment (chhagedorn) [inline src/hotspot/share/opto/mulnode.cpp:907]: While at it, you could also improve the styling here (spacing, asterisk and explicitly checking against `NULL` for `t11`).

**Review by chhagedorn — APPROVED** [review]
Looks good to me!

**Comment by TobiHartmann** [comment]
Thanks Christian, I've updated the code accordingly.

**Review by chhagedorn — APPROVED** [review]
Thanks for cleaning this up!

**Pull request closed without merging**